In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data_all = pd.read_excel('../data/Data/dataset.xlsx', names=[
    'название', 'описание'] + sum([[f'компетенция {col}', f'значимость {col}', f'содержание компетенции {col}'] \
                                   for col in range(1, 11)], []))
data_all.head(5)

название  \
0  Разработка мобильных приложений(юниоры)   
1           Преподавание в младших классах   
2            Дошкольное воспитание(юниоры)   
3                    Дошкольное воспитание   
4   Преподавание в младших классах(юниоры)   

                                            описание  \
0  Смартфоны продаются и продажи только растут. К...   
1  Высококвалифицированные специалисты в сфере на...   
2   Воспитатель детей дошкольного возраста осущес...   
3  Воспитатель детей дошкольного возраста осущест...   
4  Высококвалифицированные специалисты в сфере на...   

                                       компетенция 1 значимость 1  \
0                   Организация и управление работой            5   
1                            Общекультурное развитие           10   
2  Соблюдение санитарных норм и правил профилакти...           16   
3  Соблюдение санитарных норм и правил профилакти...           16   
4                            Общекультурное развитие           10   

                            содержание компетенции 1  \
0  Специалист должен знать и понимать: • Принципы...   
1  Специалист должен знать и понимать:\n• Основы ...   
2  Специалист должен знать и понимать:\n• Стандар...   
3  Специалист должен знать и понимать: • Стандарт...   
4  Специалист должен знать и понимать: • Основы я...   

                                       компетенция 2 значимость 2  \
0  Решение проблем, инновационность и инициативность            5   
1  Культура безопасного труда Организация рабочег...            5   
2  Поиск, анализ и оценка информации, необходимой...           24   
3  Поиск, анализ и оценка информации, необходимой...           24   
4  Культура безопасного труда Организация рабочег...            5   

                            содержание компетенции 2  \
0  Специалист должен знать и понимать: • Общие ти...   
1  Специалист должен знать и понимать: Участник к...   
2  Специалист должен знать и понимать: • основную...   
3  Специалист должен знать и понимать:\n• основну...   
4  Специалист должен знать и понимать: Участник к...   

                                       компетенция 3 значимость 3  ...  \
0      Анализ и проектирование мобильного приложения           30  ...   
1                      Общепрофессиональное развитие           40  ...   
2  Использование информационно-коммуникационных т...            6  ...   
3  Использование информационно-коммуникационных т...            6  ...   
4                      Общепрофессиональное развитие           40  ...   

  содержание компетенции 7 компетенция 8 значимость 8  \
0                      NaN           NaN          NaN   
1                      NaN           NaN          NaN   
2                      NaN           NaN          NaN   
3                      NaN           NaN          NaN   
4                      NaN           NaN          NaN   

  содержание компетенции 8 компетенция 9 значимость 9  \
0                      NaN           NaN          NaN   
1                      NaN           NaN          NaN   
2                      NaN           NaN          NaN   
3                      NaN           NaN          NaN   
4                      NaN           NaN          NaN   

  содержание компетенции 9 компетенция 10  значимость 10  \
0                      NaN            NaN            NaN   
1                      NaN            NaN            NaN   
2                      NaN            NaN            NaN   
3                      NaN            NaN            NaN   
4                      NaN            NaN            NaN   

  содержание компетенции 10  
0                       NaN  
1                       NaN  
2                       NaN  
3                       NaN  
4                       NaN  

[5 rows x 32 columns]

In [3]:
data_all.shape

(129, 32)

129 строк, в каждой из которых приведено описание компетенций, и 32 столбца, в каждом из которых приводится часть описания компетенции.

In [4]:
data_all.iloc[0]

название                               Разработка мобильных приложений(юниоры)
описание                     Смартфоны продаются и продажи только растут. К...
компетенция 1                                 Организация и управление работой
значимость 1                                                                 5
содержание компетенции 1     Специалист должен знать и понимать: • Принципы...
компетенция 2                Решение проблем, инновационность и инициативность
значимость 2                                                                 5
содержание компетенции 2     Специалист должен знать и понимать: • Общие ти...
компетенция 3                    Анализ и проектирование мобильного приложения
значимость 3                                                                30
содержание компетенции 3     Специалист должен знать и понимать: • Важность...
компетенция 4                                  Разработка мобильных приложений
значимость 4                                        

In [5]:
# пропуски
data_all.isnull().sum() / len(data_all) * 100

название                      0.000000
описание                      0.000000
компетенция 1                 0.000000
значимость 1                  0.775194
содержание компетенции 1      0.000000
компетенция 2                 0.000000
значимость 2                  0.775194
содержание компетенции 2      0.000000
компетенция 3                 0.775194
значимость 3                  0.775194
содержание компетенции 3      0.775194
компетенция 4                 3.875969
значимость 4                  3.875969
содержание компетенции 4      3.875969
компетенция 5                 8.527132
значимость 5                  9.302326
содержание компетенции 5      9.302326
компетенция 6                25.581395
значимость 6                 26.356589
содержание компетенции 6     26.356589
компетенция 7                53.488372
значимость 7                 53.488372
содержание компетенции 7     54.263566
компетенция 8                82.945736
значимость 8                 82.945736
содержание компетенции 8 

In [6]:
data_all[~data_all['компетенция 7'].isnull()]

название  \
5                Сетевое и системное администрирование   
7                        Архитектурная обработка камня   
8                       Бетонные строительные работы.    
9    ИТ-решения для бизнеса  на платформе «1С:Предп...   
10   ИТ-решения для бизнеса  на платформе «1С:Предп...   
12                           Облицовка плиткой (16-22)   
13                          Облицовка плиткой (юниоры)   
14               Информационные кабельные сети(юниоры)   
16                        Кровельные работы по металлу   
25                                     Геодезия(16-22)   
26                                    Геодезия(юниоры)   
32                                   Водные технологии   
33       Изготовление изделий из полимерных материалов   
34                     Изготовление прототипов (12-14)   
35                            Изготовление прототипов    
38                                      Интернет вещей   
39                   Командная работа на производстве    
42                   Фрезерные работы на станках с ЧПУ   
43                    Токарные работы на станках с ЧПУ   
48                  Сельскохозяйственные биотехнологии   
49                                Сварочные технологии   
51                      Промышленная механика и монтаж   
55                                  Ландшафтный дизайн   
56                         Ландшафтный дизайн (Юниоры)   
57               3D моделирование для компьютерных игр   
58                             Администрирование отеля   
59                     Администрирование отеля(Юниоры)   
60                                         Архитектура   
65                          Выпечка осетинских пирогов   
66                 Выпечка осетинских пирогов (Юниоры)   
74                       Информационные кабельные сети   
75                                   Кондитерское дело   
76                          Кондитерское дело (Юниоры)   
77   Корпоративная защита от внутренних угроз инфор...   
78                                     Кузовной ремонт   
79                            Кузовной ремонт (Юниоры)   
91                                      Поварское дело   
92                             Поварское дело (Юниоры)   
95                                 Предпринимательство   
96                        Предпринимательство (Юниоры)   
97                                 Производство мебели   
98                        Производство мебели (Юниоры)   
101                 Реставрация произведений из дерева   
102                                 Ресторанный сервис   
103                        Ресторанный сервис (Юниоры)   
106                     Сервис на воздушном транспорте   
107            Сервис на воздушном транспорте (Юниоры)   
108           Сухое строительство и штукатурные работы   
109  Сухое строительство и штукатурные работы (Юниоры)   
110                                    Технологии моды   
111                           Технологии моды (Юниоры)   
116                                         Флористика   
117                                Флористика (Юниоры)   
118                                         Фотография   
119                                Фотография (Юниоры)   
120                                       Хлебопечение   
121                              Хлебопечение (Юниоры)   
124                                      Электромонтаж   
125                             Электромонтаж (Юниоры)   
128                                     Ювелирное дело   

                                              описание  \
5    Сетевое и системное администрирование требует ...   
7    Натуральный камень – часть международного прир...   
8    Специалист по выполнению бетонных работ обычно...   
9    В современных условиях хорошая автоматизация у...   
10   В современных условиях хорошая автоматизация у...   
12    Плиточник в целом работает на коммерческих и ...   
13   Плиточник в целом работает на коммерческих и ж...   
14   Сегодня более 70% аварийных ситуаций н

In [7]:
data_all[~data_all['компетенция 10'].isnull()]

название  \
14  Информационные кабельные сети(юниоры)   
39      Командная работа на производстве    
55                     Ландшафтный дизайн   
56            Ландшафтный дизайн (Юниоры)   
95                    Предпринимательство   
96           Предпринимательство (Юниоры)   

                                             описание  \
14  Сегодня более 70% аварийных ситуаций на телеко...   
39  Соревнование по командной работе на производст...   
55  Вид соревнований: Командные соревнования, в со...   
56  Вид соревнований: Командные соревнования, в со...   
95  Это командные соревнования (в каждой команде д...   
96  Это командные соревнования (в каждой команде д...   

                                   компетенция 1 значимость 1  \
14                             Организация работ            5   
39             Организация и управление работой            10   
55  Организация и управление рабочими процессами           10   
56  Организация и управление рабочими процессами           11   
95                                   Бизнес-план            8   
96                                   Бизнес-план            8   

                             содержание компетенции 1  \
14  Специалист должен знать и понимать:  Законода...   
39  Специалист должен знать и понимать: ❖ принципы...   
55  Специалист должен знать и понимать: принципы п...   
56  Специалист должен знать и понимать: принципы п...   
95  Специалист должен знать и понимать: ● роль и з...   
96  Специалист должен знать и понимать: ● роль и з...   

                                компетенция 2 значимость 2  \
14                    Коммуникативные навыки             5   
39  Навыки общения и межличностных отношений             5   
55   Навыки коммуникации и работы с клиентами            5   
56   Навыки коммуникации и работы с клиентами            6   
95                         Организация работы            5   
96                         Организация работы            5   

                             содержание компетенции 2  \
14  Специалист должен знать и понимать:  Важность...   
39  Специалист должен знать и понимать:   ❖ спектр...   
55  Специалист должен знать и понимать: важность у...   
56  Специалист должен знать и понимать: важность у...   
95  Специалист должен знать и понимать: ● значение...   
96  Специалист должен знать и понимать: ● значение...   

                                        компетенция 3 значимость 3  ...  \
14                     Планирование и проектирование             5  ...   
39               Проектирование и реализация проекта            10  ...   
55  Проектирование зеленых насаждений и интерпрета...           15  ...   
56  Проектирование зеленых насаждений и интерпрета...           16  ...   
95  Формирование навыков коллективной работы и упр...            5  ...   
96  Формирование навыков коллективной работы и упр...            5  ...   

                             содержание компетенции 7  \
14  Специалист должен знать и понимать:  Стандарт...   
39  Специалист должен знать и понимать:   ❖ принци...   
55  Специалист должен знать и понимать:\nклассифик...   
56  Специалист должен знать и понимать:\nклассифик...   
95  Специалист должен знать и понимать: ● преимуще...   
96  Специалист должен знать и понимать: ● преимуще...   

                                        компетенция 8 значимость 8  \
14                 Поиск и устранение неисправностей          10.0   
39                              Сварочные технологии          10.0   
55  Садовые технологии (водопровод и канализация,\...          5.0   
56  Садовые технологии (водопровод и канализация,\...          6.0   
95                             Финансовые инструменты         10.0   
96                             Финансовые инструменты         10.0   

                             содержание компетенции 8  \
14  Специалист должен знать и понимать:  Этапы оп...   
39  Специалист должен знать и понимать: ❖ Специфич...   
55  Специалист должен знать и понимать

В компетенциях 7-10 приведены специфичные группы умений, которые не свойственны для других компетенций. Принимается решение удалить столбцы, связанные с этими группами компетенций.

In [8]:
data_all[~data_all['компетенция 7'].isnull()]['компетенция 7'].unique()

array(['Установка и пуско-наладка телекоммуникационного оборудования ',
       'Резьба букв, орнамента', 'Демонтаж опалубки и очистка ',
       'Документирование программных решений ', 'Закрепление',
       'Технологии «Умный дом» ', 'Всего',
       'Решение проблем и креативность ',
       'Решение проблем и креативность', 'Автоматизация и документация',
       'Изделие из полимерного материала', 'Работоспособность прототипа ',
       'Тестирование и отладка решения ', 'Электроника',
       'Обработка на станке ',
       'Завершение обработки и предоставление детали',
       'Экологические аспекты ',
       'Завершение, обеспечение качества и испытания ',
       'Техническое обслуживание, поиск неисправностей и ремонт',
       'Посадка растений и деревьев и уход за ними', 'Риггинг и анимация',
       'Управление жалобами',
       'Строительные материалы, изделия и конструкции при проектировании объектов капитального строительства или отдельных архитектурных и объемно-планировочных реш

Анкета, приведённая в примере, $-$ это образец тех данных, которые будет принимать на вход итоговая система. Нужно обратить внимание, что в анкете представлены группы навыков и показатель их важности в процентном соотношении.

Необходимо выделить такие группы в наших данных.

In [9]:
data_all['название'].unique()

array(['Разработка мобильных приложений(юниоры)',
       'Преподавание в младших классах', 'Дошкольное воспитание(юниоры)',
       'Дошкольное воспитание', 'Преподавание в младших классах(юниоры)',
       'Сетевое и системное администрирование',
       'Эксплуатация кабельных линий электропередачи',
       'Архитектурная обработка камня', 'Бетонные строительные работы. ',
       'ИТ-решения для бизнеса  на платформе «1С:Предприятие 8»',
       'ИТ-решения для бизнеса  на платформе «1С:Предприятие 8» (юниоры)',
       'Программные решения для бизнеса IT решения для бизнеса на платформе 1С Предприятие 8',
       'Облицовка плиткой (16-22)', 'Облицовка плиткой (юниоры)',
       'Информационные кабельные сети(юниоры)', 'Кровельные работы',
       'Кровельные работы по металлу', 'Печатные технологии в прессе',
       'Машинное обучение и большие данные ',
       'Машинное обучение и большие данные (Юниоры)',
       'Разработка мобильных приложений',
       'Разработка решений на базе блокче

In [10]:
# необходимо стандартизировать названия групп навыков и умений
def text_preprocessing(text):
    import re
    import string
    text = str(text)
    text = re.sub(r'\n|\r|\b', ' ', text, flags=re.S)
    text = ''.join([c for c in text if c not in string.punctuation]).lower().strip()
    text = ''.join([c for c in text if c not in string.digits])
    text = re.sub(r" +", " ", text, flags=re.S)
    return text

for i in range(1, 11):
    data_all[f'компетенция {i}'] = data_all[f'компетенция {i}'].apply(text_preprocessing)
data_all.head(5)

название  \
0  Разработка мобильных приложений(юниоры)   
1           Преподавание в младших классах   
2            Дошкольное воспитание(юниоры)   
3                    Дошкольное воспитание   
4   Преподавание в младших классах(юниоры)   

                                            описание  \
0  Смартфоны продаются и продажи только растут. К...   
1  Высококвалифицированные специалисты в сфере на...   
2   Воспитатель детей дошкольного возраста осущес...   
3  Воспитатель детей дошкольного возраста осущест...   
4  Высококвалифицированные специалисты в сфере на...   

                                       компетенция 1 значимость 1  \
0                   организация и управление работой            5   
1                            общекультурное развитие           10   
2  соблюдение санитарных норм и правил профилакти...           16   
3  соблюдение санитарных норм и правил профилакти...           16   
4                            общекультурное развитие           10   

                            содержание компетенции 1  \
0  Специалист должен знать и понимать: • Принципы...   
1  Специалист должен знать и понимать:\n• Основы ...   
2  Специалист должен знать и понимать:\n• Стандар...   
3  Специалист должен знать и понимать: • Стандарт...   
4  Специалист должен знать и понимать: • Основы я...   

                                       компетенция 2 значимость 2  \
0   решение проблем инновационность и инициативность            5   
1  культура безопасного труда организация рабочег...            5   
2  поиск анализ и оценка информации необходимой д...           24   
3  поиск анализ и оценка информации необходимой д...           24   
4  культура безопасного труда организация рабочег...            5   

                            содержание компетенции 2  \
0  Специалист должен знать и понимать: • Общие ти...   
1  Специалист должен знать и понимать: Участник к...   
2  Специалист должен знать и понимать: • основную...   
3  Специалист должен знать и понимать:\n• основну...   
4  Специалист должен знать и понимать: Участник к...   

                                       компетенция 3 значимость 3  ...  \
0      анализ и проектирование мобильного приложения           30  ...   
1                      общепрофессиональное развитие           40  ...   
2  использование информационно коммуникационных т...            6  ...   
3  использование информационно коммуникационных т...            6  ...   
4                      общепрофессиональное развитие           40  ...   

  содержание компетенции 7 компетенция 8 значимость 8  \
0                      NaN           nan          NaN   
1                      NaN           nan          NaN   
2                      NaN           nan          NaN   
3                      NaN           nan          NaN   
4                      NaN           nan          NaN   

  содержание компетенции 8 компетенция 9 значимость 9  \
0                      NaN           nan          NaN   
1                      NaN           nan          NaN   
2                      NaN           nan          NaN   
3                      NaN           nan          NaN   
4                      NaN           nan          NaN   

  содержание компетенции 9 компетенция 10  значимость 10  \
0                      NaN            nan            NaN   
1                      NaN            nan            NaN   
2                      NaN            nan            NaN   
3                      NaN            nan            NaN   
4                      NaN            nan            NaN   

  содержание компетенции 10  
0                       NaN  
1                       NaN  
2                       NaN  
3                       NaN  
4                       NaN  

[5 rows x 32 columns]

In [11]:
data_all['компетенция 3'].unique()

array(['анализ и проектирование мобильного приложения',
       'общепрофессиональное развитие',
       'использование информационно коммуникационных технологий',
       'консультирование и поддержка пользователей',
       'решение проблем инновация и креативность',
       'решение возникающих проблем', 'чтение чертежей',
       'решение проблем инновации и креативность',
       'способность решения проблем инновационный потенциал и креативность',
       'планирование и проектирование', 'монтаж слоев кровельной системы',
       'планирование и оценка',
       ' контроль качества настройка поиск и устранение copyright © союз « ворлдскиллс россия » печатные технологии в прессе г неисправностей',
       ' анализ и обработка данных ', 'программирование узла блокчейн',
       'монтаж внутреннего газопровода',
       'решение проблем инновации креативность',
       'полевые геодезические измерения с использованием современного геодезического оборудования',
       'художественный дизайн', 'uni